tests/end_to_end/test_helper.cpp

```diff
@@ -14,6 +17,97 @@
+    }
+    ~duplicate_handles_table_t() { ebpf_lock_destroy(&lock); }
+
+    bool
+    reference_or_add(ebpf_handle_t handle)
+    {
+        bool success = true;
+        auto state = ebpf_lock_lock(&lock);
+        if (!rundown_in_progress) {
+            std::map<ebpf_handle_t, uint16_t>::iterator it = table.find(handle);
+            if (it != table.end()) {
+                it->second++;
+            } else {
+                try {
+                    // The reference count of newly inserted duplicate handle is 2 (for original + first duplicate).
+                    table.insert(std::pair<ebpf_handle_t, uint16_t>(handle, static_cast<uint16_t>(2)));
+                } catch (...) {
+                    success = false;
+                }
+            }
+        }
+        ebpf_lock_unlock(&lock, state);
+        return success;
+    }
+
+    bool
+    dereference_if_found(ebpf_handle_t handle)
+    {
+        bool found = false;
+
+        auto state = ebpf_lock_lock(&lock);
+        std::map<ebpf_handle_t, uint16_t>::iterator it = table.find(handle);
+        if (it != table.end()) {
+            found = true;
+            // Dereference the handle. If the reference count drops to 0, close the handle.
+            if (--it->second == 0) {
+                table.erase(handle);
+                ebpf_api_close_handle(handle);
+            }
+            if (rundown_in_progress && table.size() == 0) {
+                // All duplicate handles have been closed. Fulfill the promise.
+                REQUIRE(all_duplicate_handles_closed != nullptr);
+                all_duplicate_handles_closed->set_value();
+            }
+        }
+
+        ebpf_lock_unlock(&lock, state);
+        return found;
+    }
+
+    void
+    rundown()
+    {
+        auto state = ebpf_lock_lock(&lock);
+        std::future<void> all_duplicate_handles_closed_callback;
+        bool duplicates_pending = false;
+        if (table.size() > 0) {
+            duplicates_pending = true;
+            all_duplicate_handles_closed = new (std::nothrow) std::promise<void>();
+            REQUIRE(all_duplicate_handles_closed != nullptr);
+            all_duplicate_handles_closed_callback = all_duplicate_handles_closed->get_future();
+            rundown_in_progress = true;
+        }
+        ebpf_lock_unlock(&lock, state);
+        if (duplicates_pending)
+            // Wait for at most 1 second for all duplicate handles to be closed.
+            REQUIRE(all_duplicate_handles_closed_callback.wait_for(1s) == std::future_status::ready);
+
+        state = ebpf_lock_lock(&lock);
+        rundown_in_progress = false;
+        delete all_duplicate_handles_closed;
+        all_duplicate_handles_closed = nullptr;
+        ebpf_lock_unlock(&lock, state);
+    }
+
+  private:
+    ebpf_lock_t lock;
+    // Map of handles to duplicate count.
+    std::map<ebpf_handle_t, uint16_t> table;
+    bool rundown_in_progress;
+    std::promise<void>* all_duplicate_handles_closed;
```

Comment:
https://github.com/microsoft/ebpf-for-windows/blob/master/docs/DevelopmentGuide.md says:

> Prefix names with _ to indicate internal and private fields or methods (e.g., _internal_field, _internal_method()).

```suggestion
    ebpf_lock_t _lock;
    // Map of handles to duplicate count.
    std::map<ebpf_handle_t, uint16_t> _table;
    bool _rundown_in_progress;
    std::promise<void>* _all_duplicate_handles_closed;
```